src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs

```diff
@@ -1058,6 +1072,29 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
             }
         }
 
+        /// <summary>
+        /// Return true if <paramref name="overridingReturnType"/> is valid for the return type of an override method when the overridden method's return type is <paramref name="overriddenReturnType"/>.
+        /// </summary>
+        private bool IsValidOverrideReturnType(Symbol overridingSymbol, TypeWithAnnotations overridingReturnType, TypeWithAnnotations overriddenReturnType, DiagnosticBag diagnostics)
```

Comment:
>IsValidOverrideReturnType [](start = 21, length = 25)

I'm curious why you structured this method this way. I would have expected:
1. check if types are equal, if so, then return true
2. check if the types are convertible, if so do the version check and report LangVer error here (so it doesn't have to be duplicated at various call sites)

The benefit of this approach is that it keeps the binding from depending on LangVer. We can verify that in testing the semantic model in error cases.

Reply:
Not all places where this is called deserve a language version diagnostic. For example, that probably isn't apropos for events.

I'm not concerned with an extra check in the case of an error, since errors such as this occur only rarely during development.

---
In reply to: [414000465](https://github.com/dotnet/roslyn/pull/43576#discussion_r414000465) [](ancestors = 414000465)

Reply:
Okay. 
I'm still concerned that we're binding differently based on LangVer and that will be visible via symbols from semantic model.

---
In reply to: [414047588](https://github.com/dotnet/roslyn/pull/43576#discussion_r414047588) [](ancestors = 414047588,414000465)

Reply:
Verifying the symbol and semantic models is a future work item.

---
In reply to: [414049357](https://github.com/dotnet/roslyn/pull/43576#discussion_r414049357) [](ancestors = 414049357,414047588,414000465)

Reply:
Sorry to insist, but I don't feel good about punting because we can see there is an issue here (should not bind differently depending on LangVer). Let's avoid introducing this problem, rather than come back and fix it later.

---
In reply to: [414052536](https://github.com/dotnet/roslyn/pull/43576#discussion_r414052536) [](ancestors = 414052536,414049357,414047588,414000465)

Reply:
This doesn't affect binding.  It only affects what diagnostics are produced. An override method with the wrong return type has always been an override of the base method. That is not changing. What is changing is when an error is produced.

---
In reply to: [414106079](https://github.com/dotnet/roslyn/pull/43576#discussion_r414106079) [](ancestors = 414106079,414052536,414049357,414047588,414000465)

Reply:
Ah, thanks! :-)